### ISSUE TITLE
Training Detail

### ISSUE THREAD

**Issue by luoyuxuan0528** [nkolot/SPIN#58]
I am try to reproducing the wonderful SPIN model from scratch.

1. The paper says  the regression model has the same design HMR with the only difference that we use the representation proposed by Zhou. I just wonder whether you train SPIN from ImageNet pre-train model or HMR pre-train model.

2. if I do not use your pre-trained SPIN checkpoint, what hyper parameters (like lr, epochs etc.) should I use to train. Is it EXACTLY the same with your training example?

3. If I train model using h3.6m without  SMPL pose&shape Moshed from 3d-joints, what's the performance in benckmark like h3.6m, 3dpw, mpi-inf-3d?

4. I cannot find the code of Mosh. If I want to perfectly reproduce the work, I have to write Mosh code my self?

p.s. it is really a nice job, Thanks a lot!

**Comment by ikvision** [comment]
About Question 1 in the code the resnet backbone is initialized from Imagenet
https://github.com/nkolot/SPIN/blob/21a148d29ea99627bd24cd51997d1484986c196e/models/hmr.py#L160-L161
The rest of the HMR layers are initialized randomly 
https://github.com/nkolot/SPIN/blob/21a148d29ea99627bd24cd51997d1484986c196e/models/hmr.py#L74-L84

**Comment by luoyuxuan0528** [comment]
Got it! any hint for other questions?

**Comment by luoyuxuan0528** [comment]
I have found much useful information in 「Training details  」of paper's supplementary  material: http://openaccess.thecvf.com/content_ICCV_2019/supplemental/Kolotouros_Learning_to_Reconstruct_ICCV_2019_supplemental.pdf

* But I still have some questions:
(1) You mention the use of openpose supervision in Main Paper, but the weight of that is set to 0.0 in the code & nothing is mentioned in Supp. Mat. what's the correct value of --gt_train_weight and --openpose_train_weight
(2) value of --smplify_threshold is not mention in paper, is 100 good enough?
(3) The paired model is trained "on Human3.6M [4] using full 3D pose and shape ground truth". What's the "shape" refer to, the beta loss of SMPL or the Mesh level loss?
(4) what's the hyper-parameter to train the pre-trained model, could you release it as well?

Looking for your reply.

P.S. there all some small bug and mis-alignment with paper I found in utils/training_opt.py.

train.add_argument('--openpose_train_weight', default=0., help='Weight for OpenPose keypoints during training')  **type=float is missing**

 train.add_argument('--num_smplify_iters', default=100, type=int, help='Number of SMPLify iterations') **value in paper is 50**

**Comment by nkolot** [comment]
Sorry for the late reply but we've been very busy.

Wherever there is a difference between the defaults in the released code and the values mentioned in the paper/supp-mat I recommend using the ones reported in the paper.


1) We do not train with OpenPose so I suggest setting this to 0. This constant is used because of the way the code is structured.
2) Yes you should use 100.
3) Shape refers to the SMPL betas.
4) Using the same hyperparameters should do the job.

**Comment by luoyuxuan0528** [comment]
Thanks for that! I have trained a model with 41.8mm on h36m. (the same as yours eval on my machine.)
But performance on 3dpw is not good (82.2 mm reconstruction error).  
Any Suggestion ?

**Comment by nkolot** [comment]
Did you fix this issue? 82.2mm is too high there is definitely something wrong here.
What is your training process?

**Comment by luoyuxuan0528** [comment]
Not yet.  My Training process:

(1) training on h36m about 60epoch got a PAMPJPE 52mm(on h36m) model. (dataset_ratio is a new arg I added)

    python train.py --name h36m_pretrain_trial_4 --num_epochs 63 --lr 3e-5 --keypoint_loss_weight 5.0 --openpose_train_weight 0.0 --gt_train_weight 1.0 --dataset_ratio 1.0 0.0 0.0 --summary_steps 100

(2) training on all datasets 60epoch got a PAPMPJPE 42mm(on h36m) model, but performance on 3dpw is 82.2.
    python train.py --name h36m_pretrain_trial_4_finetune_bs128_0.5_0.5 --num_epochs 62 --lr 6e-5 --keypoint_loss_weight 5.0 --openpose_train_weight 0.5 --gt_train_weight 0.5 --dataset_ratio 0.3 0.6 0.1 --summary_steps 500 --eval_datasets h36m-p2 --run_smplify --num_smplify_iters 50 --test_steps 5000 --checkpoint_steps 5000 --batch_size 128 --pretrained_checkpoint xxx

I've try to train another backbone, the result is similar. If I use your final fits instead of --run_simpify , the result is much better, about 59~60mm on 3dpw.

**Comment by nkolot** [comment]
The number that you get by using the final fits is similar to the one reported in the paper and this is what you should expect.
In the paper we mention that we enable the in-the-loop optimization after pertaining the model first. Did you try to pretrain the model with the original static fits for e.g. 30 epochs?

**Comment by nkolot** [comment]
I suggested this as a debugging strategy to see if it works better, since it worked for the unpaired case.
How does the optimized shape look during training? Does it seem to converge to a good solution?

**Comment by luoyuxuan0528** [comment]
The optimized shape in tensorboard looks good, I will try your suggestion.

**Comment by andgitchang** [comment]
Hi @luoyuxuan0528 , I am interested in your third question:

> 3. If I train model using h3.6m without SMPL pose&shape Moshed from 3d-joints, what's the performance in benckmark like h3.6m, 3dpw, mpi-inf-3d?

Did the results of 41.8mm on h36m / 82.2 mm on 3dpw be trained without Moshed Human3.6? Otherwise where did you get Mosh by yourself? Thank you.

**Comment by sparshgarg23** [comment]
Hello, luoyuxyan0528,How did you access the final fits.I am trying to download them with the link but it seems that the link is no longer working.
Also,after training the model for 30 epochs with the static fits,did you use fine tuning for the remainder 30 epochs with in loop optimization?

**Comment by sparshgarg23** [comment]
@ikvision  having issues with SPIN/HMR.py when it loads the pretrained resnet50 model,it gives me a size mismatch error of the following form
```
rror(s) in loading state_dict for Predictor:
	size mismatch for conv1.weight: copying a param with shape torch.Size([64, 3, 7, 7]) from checkpoint, the shape in current model is torch.Size([64, 3, 3, 3]).
	size mismatch for layer1.0.conv2.weight: copying a param with shape torch.Size([64, 64, 3, 3]) from checkpoint, the shape in current model is torch.Size([64, 64, 7, 7]).
	size mismatch for layer1.0.bn2.weight: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.0.bn2.bias: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.0.bn2.running_mean: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.0.bn2.running_var: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.1.conv2.weight: copying a param with shape torch.Size([64, 64, 3, 3]) from checkpoint, the shape in current model is torch.Size([64, 64, 7, 7]).
	size mismatch for layer1.1.bn2.weight: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.1.bn2.bias: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.1.bn2.running_mean: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.1.bn2.running_var: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.2.conv2.weight: copying a param with shape torch.Size([64, 64, 3, 3]) from checkpoint, the shape in current model is torch.Size([64, 64, 7, 7]).
	size mismatch for layer1.2.bn2.weight: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.2.bn2.bias: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.2.bn2.running_mean: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer1.2.bn2.running_var: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([256]).
	size mismatch for layer2.0.conv2.weight: copying a param with shape torch.Size([128, 128, 3, 3]) from checkpoint, the shape in current model is torch.Size([128, 128, 7, 7]).
	size mismatch for layer2.0.bn2.weight: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.0.bn2.bias: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.0.bn2.running_mean: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.0.bn2.running_var: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.1.conv2.weight: copying a param with shape torch.Size([128, 128, 3, 3]) from checkpoint, the shape in current model is torch.Size([128, 128, 7, 7]).
	size mismatch for layer2.1.bn2.weight: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.1.bn2.bias: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.1.bn2.running_mean: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.1.bn2.running_var: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.2.conv2.weight: copying a param with shape torch.Size([128, 128, 3, 3]) from checkpoint, the shape in current model is torch.Size([128, 128, 7, 7]).
	size mismatch for layer2.2.bn2.weight: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.2.bn2.bias: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.2.bn2.running_mean: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.2.bn2.running_var: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.3.conv2.weight: copying a param with shape torch.Size([128, 128, 3, 3]) from checkpoint, the shape in current model is torch.Size([128, 128, 7, 7]).
	size mismatch for layer2.3.bn2.weight: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.3.bn2.bias: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.3.bn2.running_mean: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer2.3.bn2.running_var: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([512]).
	size mismatch for layer3.0.conv2.weight: copying a param with shape torch.Size([256, 256, 3, 3]) from checkpoint, the shape in current model is torch.Size([256, 256, 7, 7]).
	size mismatch for layer3.0.bn2.weight: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.0.bn2.bias: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.0.bn2.running_mean: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.0.bn2.running_var: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.1.conv2.weight: copying a param with shape torch.Size([256, 256, 3, 3]) from checkpoint, the shape in current model is torch.Size([256, 256, 7, 7]).
	size mismatch for layer3.1.bn2.weight: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.1.bn2.bias: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.1.bn2.running_mean: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.1.bn2.running_var: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.2.conv2.weight: copying a param with shape torch.Size([256, 256, 3, 3]) from checkpoint, the shape in current model is torch.Size([256, 256, 7, 7]).
	size mismatch for layer3.2.bn2.weight: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.2.bn2.bias: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.2.bn2.running_mean: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.2.bn2.running_var: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.3.conv2.weight: copying a param with shape torch.Size([256, 256, 3, 3]) from checkpoint, the shape in current model is torch.Size([256, 256, 7, 7]).
	size mismatch for layer3.3.bn2.weight: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.3.bn2.bias: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.3.bn2.running_mean: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.3.bn2.running_var: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.4.conv2.weight: copying a param with shape torch.Size([256, 256, 3, 3]) from checkpoint, the shape in current model is torch.Size([256, 256, 7, 7]).
	size mismatch for layer3.4.bn2.weight: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.4.bn2.bias: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.4.bn2.running_mean: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.4.bn2.running_var: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.5.conv2.weight: copying a param with shape torch.Size([256, 256, 3, 3]) from checkpoint, the shape in current model is torch.Size([256, 256, 7, 7]).
	size mismatch for layer3.5.bn2.weight: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.5.bn2.bias: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.5.bn2.running_mean: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer3.5.bn2.running_var: copying a param with shape torch.Size([256]) from checkpoint, the shape in current model is torch.Size([1024]).
	size mismatch for layer4.0.conv2.weight: copying a param with shape torch.Size([512, 512, 3, 3]) from checkpoint, the shape in current model is torch.Size([512, 512, 7, 7]).
	size mismatch for layer4.0.bn2.weight: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.0.bn2.bias: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.0.bn2.running_mean: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.0.bn2.running_var: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.1.conv2.weight: copying a param with shape torch.Size([512, 512, 3, 3]) from checkpoint, the shape in current model is torch.Size([512, 512, 7, 7]).
	size mismatch for layer4.1.bn2.weight: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.1.bn2.bias: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.1.bn2.running_mean: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.1.bn2.running_var: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.2.conv2.weight: copying a param with shape torch.Size([512, 512, 3, 3]) from checkpoint, the shape in current model is torch.Size([512, 512, 7, 7]).
	size mismatch for layer4.2.bn2.weight: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.2.bn2.bias: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.2.bn2.running_mean: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
	size mismatch for layer4.2.bn2.running_var: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([2048]).
```
